src/libpmem/pmem.c

```diff
@@ -1282,3 +1589,70 @@ pmem_init(void)
 			"QueryVirtualMemoryInformation");
 #endif
 }
+
+
+/*
+ * range_deep_flush -- (internal) perform deep flush of given address range
+ */
+static int
+range_deep_flush(uintptr_t addr, size_t len)
+{
+
+	while (len != 0) {
+		const struct map_tracker *mt = map_range_find(addr, len);
+
+		if (mt == NULL) /* no more overlapping track regions */
+			return pmem_msync((void *)addr, len);
+
+		if (mt->base_addr > addr) {
+			size_t curr_len = mt->base_addr - addr;
+			if (curr_len > len)
+				curr_len = len;
+			if (pmem_msync((void *)addr, curr_len) != 0)
+				return -1;
+			if ((len -= curr_len) == 0)
+				return 0;
+			addr = mt->base_addr;
+		}
+
+		if (ddax_deep_flush_write(mt->region_id) < 0) {
+			ERR("!cannot perform final write to deep_flush");
```

Comment:
"!" should be used only for lowest level functions. Also please print region_id.